File: cpp/utils/disjoint_set.cpp
#include "disjoint_set.hpp"

DisjointSet::DisjointSet(int n) {
  ds = new int[n];
  for (int i = 0; i < n; i++) {
    ds[i] = i;
  }
}
DisjointSet::~DisjointSet() { delete[] ds; }

int DisjointSet::Find(int x) {
  int parent = ds[x];
  if (parent == x) {
    return x;
  }
  return Find(parent);
}

void DisjointSet::Union(int x, int y) {
  int px = Find(x);
  int py = Find(y);
  if (px < py) {
    ds[py] = px;
  } else {
    ds[px] = py;
  }
}

bool DisjointSet::Check(int x, int y) { return Find(x) == Find(y); }

File: go/data-structures/set/disjoint_set.go
package set

type DisjointSet []int

func NewDisjointSet(n int) DisjointSet {
	ds := []int{}
	for i := range n {
		ds = append(ds, i)
	}
	return ds
}

func (ds DisjointSet) Find(x int) int {
	parent := ds[x]
	if parent == x {
		return x
	}
	return ds.Find(parent)
}

func (ds DisjointSet) Union(x, y int) {
	px := ds.Find(x)
	py := ds.Find(y)
	if px < py {
		ds[py] = px
	} else {
		ds[px] = py
	}
}

func (ds DisjointSet) Check(x, y int) bool {
	return ds.Find(x) == ds.Find(y)
}
